File: examples/go/packages/data/json_file_store.go
package data

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

type JSONFileStore[T any] struct {
	filePath     string
	fallbackData T
	mu           sync.Mutex
}

func NewJSONFileStore[T any](filePath string, fallbackData T) *JSONFileStore[T] {
	return &JSONFileStore[T]{
		filePath:     filePath,
		fallbackData: fallbackData,
	}
}

func (s *JSONFileStore[T]) FilePath() string {
	return s.filePath
}

func (s *JSONFileStore[T]) Read() (T, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	var value T
	err := s.withFileLock(func() error {
		if err := s.ensureFileExists(); err != nil {
			return err
		}

		readValue, err := s.readJSONFile()
		if err != nil {
			return err
		}

		value = readValue
		return nil
	})

	return value, err
}

func (s *JSONFileStore[T]) Write(value T) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	return s.withFileLock(func() error {
		if err := s.ensureFileExists(); err != nil {
			return err
		}

		return s.writeJSONAtomically(value)
	})
}

func (s *JSONFileStore[T]) Reset() error {
	s.mu.Lock()
	defer s.mu.Unlock()

	return s.withFileLock(func() error {
		if err := s.ensureFileExists(); err != nil {
			return err
		}

		return s.writeJSONAtomically(s.fallbackData)
	})
}

func (s *JSONFileStore[T]) Update(updater func(current T) (T, error)) (T, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	var next T
	err := s.withFileLock(func() error {
		if err := s.ensureFileExists(); err != nil {
			return err
		}

		current, err := s.readJSONFile()
		if err != nil {
			return err
		}

		updated, err := updater(current)
		if err != nil {
			return err
		}

		next = updated
		return s.writeJSONAtomically(updated)
	})

	return next, err
}

func (s *JSONFileStore[T]) ensureFileExists() error {
	if err := os.MkdirAll(filepath.Dir(s.filePath), 0o755); err != nil {
		return err
	}

	_, err := os.Stat(s.filePath)
	if err == nil {
		return nil
	}

	if !errors.Is(err, os.ErrNotExist) {
		return err
	}

	return s.writeJSONAtomically(s.fallbackData)
}

func (s *JSONFileStore[T]) readJSONFile() (T, error) {
	var value T

	content, err := os.ReadFile(s.filePath)
	if err != nil {
		return value, err
	}

	if err := json.Unmarshal(content, &value); err != nil {
		return value, err
	}

	return value, nil
}

func (s *JSONFileStore[T]) writeJSONAtomically(value T) error {
	content, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		return err
	}

	if err := os.MkdirAll(filepath.Dir(s.filePath), 0o755); err != nil {
		return err
	}

	tempFilePath := fmt.Sprintf(
		"%s.%d.%d.tmp",
		s.filePath,
		os.Getpid(),
		time.Now().UnixNano(),
	)

	if err := os.WriteFile(tempFilePath, append(content, '\n'), 0o644); err != nil {
		return err
	}

	if err := os.Rename(tempFilePath, s.filePath); err != nil {
		_ = os.Remove(tempFilePath)
		return err
	}

	return nil
}

func (s *JSONFileStore[T]) withFileLock(operation func() error) error {
	lockPath := s.filePath + ".lock"
	timeout := 5 * time.Second
	retryDelay := 25 * time.Millisecond
	staleLockAge := 30 * time.Second
	startedAt := time.Now()

	for {
		if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
			return err
		}

		lockFile, err := os.OpenFile(lockPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o644)
		if err != nil {
			if !errors.Is(err, os.ErrExist) {
				return err
			}

			if cleanupErr := cleanupStaleLock(lockPath, staleLockAge); cleanupErr != nil {
				return cleanupErr
			}

			if time.Since(startedAt) >= timeout {
				return fmt.Errorf("timed out acquiring JSONFileStore lock for %s", s.filePath)
			}

			time.Sleep(retryDelay)
			continue
		}

		_, writeErr := lockFile.WriteString(
			fmt.Sprintf("{\"pid\":%d,\"acquiredAt\":%q}\n", os.Getpid(), time.Now().UTC().Format(time.RFC3339Nano)),
		)
		closeErr := lockFile.Close()
		if writeErr != nil {
			_ = os.Remove(lockPath)
			return writeErr
		}
		if closeErr != nil {
			_ = os.Remove(lockPath)
			return closeErr
		}

		defer func() {
			_ = os.Remove(lockPath)
		}()

		return operation()
	}
}

func cleanupStaleLock(lockPath string, maxAge time.Duration) error {
	info, err := os.Stat(lockPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil
		}
		return err
	}

	if time.Since(info.ModTime()) <= maxAge {
		return nil
	}

	if err := os.Remove(lockPath); err != nil && !errors.Is(err, os.ErrNotExist) {
		return err
	}

	return nil
}

File: examples/python/packages/data/json_file_store.py
# ==========================================================================
# APP v1.1.5
# packages/data/json_file_store.py
# --------------------------------------------------------------------------
# Thread-safe, file-locked JSON store for local persistence.
#
# Features:
# - In-process serialization via threading.Lock
# - File-based lock via fcntl.flock for cross-process safety
# - Atomic writes via temp file + os.rename
# - Stale lock cleanup (30s timeout)
# - read(), write(), reset(), update() operations
#
# This is a shared package — exposed to Cases via ctx.packages,
# never imported directly by Cases.
# ==========================================================================

from __future__ import annotations

import fcntl
import json
import os
import tempfile
import threading
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Generic, TypeVar

T = TypeVar("T")

_LOCK_TIMEOUT_S = 5.0
_LOCK_RETRY_MS = 25
_STALE_LOCK_S = 30.0


@dataclass
class JsonFileStoreOptions(Generic[T]):
    """Options for creating a JsonFileStore."""
    file_path: str
    fallback_data: T


class JsonFileStore(Generic[T]):
    """
    Thread-safe, file-locked JSON file store.

    Provides serialized access to a JSON file with:
    - in-process threading.Lock for same-process concurrency
    - fcntl.flock for cross-process concurrency
    - atomic writes via temp file + os.rename
    """

    def __init__(self, options: JsonFileStoreOptions[T]) -> None:
        self._file_path = options.file_path
        self._fallback_data = options.fallback_data
        self._lock = threading.Lock()
        self._ensure_file_exists()

    @property
    def file_path(self) -> str:
        """Return the path to the underlying JSON file."""
        return self._file_path

    def read(self) -> T:
        """Read and return the current data from the file."""
        with self._lock:
            return self._locked_read()

    def write(self, data: T) -> None:
        """Overwrite the file with new data."""
        with self._lock:
            self._locked_write(data)

    def reset(self) -> None:
        """Reset the file to its fallback data."""
        with self._lock:
            self._locked_write(self._fallback_data)

    def update(self, updater: Callable[[T], T]) -> T:
        """
        Atomically read, transform, and write data.

        The updater function receives the current data and must return
        the new data. The entire operation is serialized.
        """
        with self._lock:
            current = self._locked_read()
            updated = updater(current)
            self._locked_write(updated)
            return updated

    # ==================================================================
    # Internal file operations (must be called under self._lock)
    # ==================================================================

    def _locked_read(self) -> T:
        """Read the file contents. Caller must hold self._lock."""
        lock_fd = self._acquire_file_lock()
        try:
            with open(self._file_path, "r", encoding="utf-8") as f:
                content = f.read()
            if not content.strip():
                return self._fallback_data
            return json.loads(content)  # type: ignore[no-any-return]
        except (json.JSONDecodeError, FileNotFoundError):
            return self._fallback_data
        finally:
            self._release_file_lock(lock_fd)

    def _locked_write(self, data: T) -> None:
        """Write data atomically. Caller must hold self._lock."""
        lock_fd = self._acquire_file_lock()
        try:
            dir_name = os.path.dirname(self._file_path) or "."
            fd, tmp_path = tempfile.mkstemp(dir=dir_name, suffix=".tmp")
            try:
                with os.fdopen(fd, "w", encoding="utf-8") as f:
                    json.dump(data, f, indent=2, ensure_ascii=False)
                os.rename(tmp_path, self._file_path)
            except BaseException:
                # Clean up temp file on failure.
                try:
                    os.unlink(tmp_path)
                except OSError:
                    pass
                raise
        finally:
            self._release_file_lock(lock_fd)

    # ==================================================================
    # File locking via fcntl.flock
    # ==================================================================

    def _acquire_file_lock(self) -> int:
        """
        Acquire an exclusive file lock with timeout and stale lock cleanup.

        Returns the lock file descriptor.
        """
        lock_path = self._file_path + ".lock"
        deadline = time.monotonic() + _LOCK_TIMEOUT_S

        while True:
            try:
                # Try to clean up stale locks.
                self._cleanup_stale_lock(lock_path)

                fd = os.open(lock_path, os.O_CREAT | os.O_WRONLY)
                try:
                    fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
                    return fd
                except (OSError, BlockingIOError):
                    os.close(fd)
            except OSError:
                pass

            if time.monotonic() >= deadline:
                raise TimeoutError(
                    f"JsonFileStore: could not acquire lock for {self._file_path} "
                    f"within {_LOCK_TIMEOUT_S}s"
                )

            time.sleep(_LOCK_RETRY_MS / 1000.0)

    def _release_file_lock(self, fd: int) -> None:
        """Release the file lock and close the descriptor."""
        try:
            fcntl.flock(fd, fcntl.LOCK_UN)
        finally:
            try:
                os.close(fd)
            except OSError:
                pass

    def _cleanup_stale_lock(self, lock_path: str) -> None:
        """Remove a lock file if it is older than the stale threshold."""
        try:
            stat = os.stat(lock_path)
            age = time.time() - stat.st_mtime
            if age > _STALE_LOCK_S:
                os.unlink(lock_path)
        except OSError:
            pass

    # ==================================================================
    # Initialization
    # ==================================================================

    def _ensure_file_exists(self) -> None:
        """Create the file with fallback data if it does not exist."""
        dir_name = os.path.dirname(self._file_path)
        if dir_name:
            os.makedirs(dir_name, exist_ok=True)

        if not os.path.exists(self._file_path):
            with open(self._file_path, "w", encoding="utf-8") as f:
                json.dump(self._fallback_data, f, indent=2, ensure_ascii=False)


def create_json_file_store(file_path: str, fallback_data: T) -> JsonFileStore[T]:
    """Factory function to create a JsonFileStore."""
    return JsonFileStore(JsonFileStoreOptions(file_path=file_path, fallback_data=fallback_data))
